### PR TITLE
feat(search): Similarity search for Opensearch and Elasticsearch

### PR DESCRIPTION
**Pull request by frascuchon**
# Description

This PR includes a bunch of changes regarding the `search_engine` module: 

- The base class includes a new method to tackle vector generation and defines the `similarity_search` method.
- Two similar implementations of the `SearchEngine` class have been included `OpenSearchEngine` and `ElasticSearchEngine`. Since they share a lot of functionality, a common abstract class has been implemented for common code blocks, letting each implementation handle its own details.
- A new way for engine registration and generation has been developed. This new way is based on a class decorator. The aim of this approach is to simplify future engine implementations, keeping the code isolated for each new implementation.
The following example shows how a new engine could be implemented and registered as an available engine for the Argilla server:
```python
from argilla.server.search_engine import SearchEngine

@SearchEngine.register(engine_name="my-new-engine")
class NewSearchEngineImplementation(SearchEngine):
   # implemente all required method
   ...
```
Then, you can use your search engine by environment variable:
```bash
ARGILLA_SEARCH_ENGINE=my-new-engine python -m argilla server
```
The `ARGILLA_SEARCH_ENGINE` environment variable has been added as a common server setting. This is a first approach and may change in the future. This is the reason why I did not include it in the docs section.

Refs #3028

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [X] New feature (non-breaking change which adds functionality)
- [X] Refactor (change restructuring the codebase without changing functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla/server/search_engine/elasticsearch.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+from argilla.server.search_engine import (
+    SearchEngine,
+    SearchResponses,
+    UserResponseStatusFilter,
+)
+from argilla.server.search_engine.commons import BaseElasticAndOpenSearchEngine, field_name_for_vector_settings
+from argilla.server.settings import settings
+
+
+def _compute_num_candidates_from_k(k: int):
```
Comment (gabrielmbmb) [inline src/argilla/server/search_engine/elasticsearch.py]:
```suggestion
def _compute_num_candidates_from_k(k: int) -> int:
```

File: src/argilla/server/search_engine/elasticsearch.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+class ElasticSearchEngine(BaseElasticAndOpenSearchEngine):
+    config: Dict[str, Any]
+
+    es_number_of_shards: int
+    es_number_of_replicas: int
+
+    def __post_init__(self):
+        self.client = AsyncElasticsearch(**self.config)
+
+    def _configure_index_settings(self):
```
Comment (gabrielmbmb) [inline src/argilla/server/search_engine/elasticsearch.py]:
```suggestion
    def _configure_index_settings(self) -> Dict[str, Any]:
```

File: src/argilla/server/search_engine/elasticsearch.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+    def _mapping_for_vector_settings(self, vector_settings: VectorSettings) -> dict:
+        return {
+            f"vectors.{vector_settings.id}": {
+                "type": "dense_vector",
+                "dims": vector_settings.dimensions,
+                "index": True,
+                # can similarity property also be part of config @frascuchon ?
+                # relates vector search similarity metric
+                "similarity": "l2_norm",  ## default value regarding the knn best practices es documentation
```
Comment (gabrielmbmb) [inline src/argilla/server/search_engine/elasticsearch.py:80]: I think that we should let the user configure this when creating settings for a collection of vectors. In some cases, depending on the vectors used, one metric could be better than other.

**Review by gabrielmbmb — COMMENTED** [review]
I think that we should allow the user to configure the similarity metric. Also, I think we could assume that no matter the vector database that Argilla is using, HNSW algorithm will be used.

HNSW is the default ANN algorithm in all the vector databases, and I think it's important also to let the user configure the parameters of the algorithm `ef`, `m` and `ef_construction`, as these would affect the results of the search.

File: src/argilla/server/search_engine/elasticsearch.py (hunk 3 above)
Comment (frascuchon) [reply src/argilla/server/search_engine/elasticsearch.py:80]: Yes. We can iterate over it in future PRs. The `similarity` param could affect other default ones.

File: src/argilla/server/search_engine/base.py [hunk 4]
```diff
@@ -96,3 +96,24 @@ async def search(
...
+
+    @abstractmethod
+    async def similarity_search(
+        self,
+        dataset: Dataset,
+        vector_settings: VectorSettings,
+        value: Optional[List[float]] = None,
+        record: Optional[Record] = None,
+        user_response_status_filter: Optional[UserResponseStatusFilter] = None,
+        max_results: conint(ge=2, le=500) = 100,
```
Comment (gabrielmbmb) [inline src/argilla/server/search_engine/base.py]: No validation will be applied here by `pydantic`, so let's just use `int`

```suggestion
        max_results: int = 100,
```

File: .github/workflows/run-python-tests.yml [hunk 5]
```diff
@@ -77,6 +77,16 @@ jobs:
         env:
           HF_HUB_ACCESS_TOKEN: ${{ secrets.HF_HUB_ACCESS_TOKEN }}
         run: echo "Enable HF access token"
+      - name: Set huggingface hub credentials
```
Comment (gabrielmbmb) [inline .github/workflows/run-python-tests.yml]:
```suggestion
      - name: Set Argilla search engine env variable
```

File: .github/workflows/run-python-tests.yml [hunk 6]
```diff
@@ -77,6 +77,16 @@ jobs:
...
+        if: startsWith(searchEngineDockerImage, 'docker.elastic.co')
+        env:
+          ARGILLA_SEARCH_ENGINE: elasticsearch
+        run: echo "Configure elasticsearch engine"
+      - name: Set huggingface hub credentials
```
Comment (gabrielmbmb) [inline .github/workflows/run-python-tests.yml]:
```suggestion
      - name: Set Argilla search engine env variable
```

File: src/argilla/server/search_engine/elasticsearch.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            verify_certs=settings.elasticsearch_ssl_verify,
+            ca_certs=settings.elasticsearch_ca_path,
+            retry_on_timeout=True,
+            max_retries=5,
+        )
+        return ElasticSearchEngine(
+            config,
+            es_number_of_shards=settings.es_records_index_shards,
+            es_number_of_replicas=settings.es_records_index_replicas,
+        )
```
Comment (gabrielmbmb) [inline src/argilla/server/search_engine/elasticsearch.py]:
```suggestion
    @classmethod
    async def new_instance(cls) -> "ElasticSearchEngine":
        config = dict(
            hosts=settings.elasticsearch,
            verify_certs=settings.elasticsearch_ssl_verify,
            ca_certs=settings.elasticsearch_ca_path,
            retry_on_timeout=True,
            max_retries=5,
        )
        return cls(
            config,
            es_number_of_shards=settings.es_records_index_shards,
            es_number_of_replicas=settings.es_records_index_replicas,
        )
```

File: src/argilla/server/search_engine/elasticsearch.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+def _compute_num_candidates_from_k(k: int) -> int:
+    if k < 50:
+        return 500
+    elif 50 <= k < 200:
+        return 100
+    return 2000
+
+
+@SearchEngine.register(engine_name="opensearch")
```
Comment (gabrielmbmb) [inline src/argilla/server/search_engine/elasticsearch.py]:
```suggestion
@SearchEngine.register(engine_name="elasticsearch")
```

**Review by gabrielmbmb — APPROVED** [review]

**Pull request merged**